.github/scripts/add_release_label_to_pr_and_linked_issues.ts

```diff
@@ -0,0 +1,370 @@
+                  ... on Issue {
+                    number
+                    id
+                    repository {
+                      name
+                      owner {
+                        login
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  `;
+
+  const retrieveTimelineEventsResult: {
+    repository: {
+      pullRequest: {
+        timelineItems: {
+          nodes: Array<{
+            __typename: 'ConnectedEvent' | 'DisconnectedEvent';
+            createdAt: string;
+            subject: {
+              number: number;
+              id: string;
+              repository: {
+                name: string;
+                owner: {
+                  login: string;
+                };
+              };
+            };
+          }>;
+        };
+      };
+    };
+  } = await octokit.graphql(retrieveTimelineEventsQuery, {
+    repoOwner,
+    repoName,
+    prNumber,
+  });
+
+  const timelineEvents = retrieveTimelineEventsResult?.repository?.pullRequest?.timelineItems?.nodes;
+
+  return timelineEvents;
+}
+
+// This function retrieves the list of linked issues for a pull request
+async function retrieveLinkedIssues(octokit: InstanceType<typeof GitHub>, repoOwner: string, repoName: string, prNumber: number): Promise<Labelable[]> {
+
+  // The list of linked issues can be deduced from timeline events
+  const timelineEvents = await retrieveTimelineEvents(octokit, repoOwner, repoName, prNumber);
+
+  const linkedIssuesMap: Record<string, Labelable> = {};
+
+  // This way to retrieve linked issues is not straightforward, but there's currently no easier way to obtain linked issues thanks to Github APIs
+  timelineEvents?.forEach((event) => {
+    const issue = event.subject;
+
+
+    if (event?.__typename === 'ConnectedEvent') {
+      linkedIssuesMap[issue.id] = {
+        id: issue.id,
+        number: issue?.number,
+        repoOwner: issue?.repository?.owner?.login,
+        repoName: issue?.repository?.name,
+        createdAt: event?.createdAt,
+      };
+    } else if (event?.__typename === 'DisconnectedEvent') {
+      const linkedIssue = linkedIssuesMap[issue.id] as Labelable;
+
+      if (linkedIssue && new Date(event?.createdAt) > new Date(linkedIssue?.createdAt)) {
+        delete linkedIssuesMap[issue.id];
+      }
+    }
+  });
+
+  const linkedIssues = Object.values(linkedIssuesMap);
+
+  return linkedIssues;
+}
+
+// This function adds label to a labelable object (i.e. a pull request or an issue)
+async function addLabelToLabelable(octokit: InstanceType<typeof GitHub>, labelable: Labelable, labelName: string, labelColor: string, labelDescription: string): Promise<void> {
+
+  // Retrieve label from the labelable's repo, or create label if required
+  const labelId = await createOrRetrieveLabel(octokit, labelable?.repoOwner, labelable?.repoName, labelName, labelColor, labelDescription);
```

Comment:
Nit: perhaps it would be more efficient to move this into the `main` function, so we only retrieve the label id once. At the moment we seem to be retrieving it once per issue

Reply:
The problem is `labelId` depends on  `repoOwner` and `repoName` which are potentially different for each labelable (i.e. linked issue).

Reply:
Ahhh good point, I didn't realize that, thanks